File: Level_Hard/Sincronizar_Red_Cuantica.py
#Una red cuántica tiene n nodos con estados binarios (0/1). Cada segundo, los nodos 
# actualizan su estado:
#Si un nodo tiene ≥2 vecinos con estado 1, cambia su estado (0→1 o 1→0).
#Excepción: Nodos con imán cuántico (lista M) nunca cambian.
#Determina en cuántos segundos la red alcanza un estado estable (que no cambia). Si es cíclica, retorna -1.
#Entrada:
#n m: nodos y aristas (1 ≤ n ≤ 20)
#m líneas: aristas u v (no dirigidas)
#estado_inicial: cadena binaria de longitud n
#k magnetizados: número de nodos magnetizados y sus índices
#Salida:
#Número de segundos hasta estabilidad (o -1 si es cíclica).

#Solución (Simulación + Detección de Ciclos):
from collections import deque

def main():
    import sys
    data = sys.stdin.read().splitlines()
    n, m = map(int, data[0].split())
    graph = [[] for _ in range(n)]
    for i in range(1, 1+m):
        u, v = map(int, data[i].split())
        graph[u].append(v)
        graph[v].append(u)
    
    estado = list(map(int, list(data[1+m])))
    magnetizados = set(map(int, data[2+m].split()[1:]))
    
    visited = {}
    steps = 0
    current = tuple(estado)
    
    while current not in visited:
        visited[current] = steps
        next_estado = list(current)
        changed = False
        
        for nodo in range(n):
            if nodo in magnetizados:
                continue
            vecinos = graph[nodo]
            count_1 = sum(current[v] for v in vecinos)
            if count_1 >= 2:
                next_estado[nodo] = 1 - current[nodo]
                changed = True
        
        if not changed:
            print(steps)
            return
        
        current = tuple(next_estado)
        steps += 1
    
    print(-1)

if __name__ == "__main__":
    main()

#Entrada:
#3 3
#0 1
#1 2
#2 0
#010
#1 1

#Salida: 2

File: Level_Hard/Sincronizar_Red_Cuantica.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
using namespace std;

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> graph(n);
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    
    string estado_str;
    cin >> estado_str;
    vector<int> estado(n);
    for (int i = 0; i < n; i++) {
        estado[i] = estado_str[i] - '0';
    }
    
    int k;
    cin >> k;
    set<int> magnetizados;
    for (int i = 0; i < k; i++) {
        int node;
        cin >> node;
        magnetizados.insert(node);
    }
    
    map<vector<int>, int> visited;
    int steps = 0;
    
    while (visited.find(estado) == visited.end()) {
        visited[estado] = steps;
        vector<int> next_estado = estado;
        bool changed = false;
        
        for (int nodo = 0; nodo < n; nodo++) {
            if (magnetizados.find(nodo) != magnetizados.end()) continue;
            
            int count_1 = 0;
            for (int vecino : graph[nodo]) {
                if (estado[vecino] == 1) count_1++;
            }
            
            if (count_1 >= 2) {
                next_estado[nodo] = 1 - estado[nodo];
                changed = true;
            }
        }
        
        if (!changed) {
            cout << steps << endl;
            return 0;
        }
        
        estado = next_estado;
        steps++;
    }
    
    cout << -1 << endl;
    return 0;
}